### PR TITLE
Sign up

### PR DESCRIPTION
**Pull request by edwardGr**

### CODE REVIEW

File: src/main/java/com/george/messenger/web/response/SignUpResponse.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+public class SignUpResponse extends AbstractResponse {
+
+    @JsonInclude(JsonInclude.Include.NON_EMPTY)
+    private Map<String, String> errors = new HashMap<>();
+
+    @JsonInclude(JsonInclude.Include.NON_EMPTY)
+    private String errorMessage;
+
+    @JsonIgnore
+    private URI createdResourceUri;
```
Comment (VadymPolyanski) [inline src/main/java/com/george/messenger/web/response/SignUpResponse.java]: URI? errors?
maybe just token?

File: src/main/java/com/george/messenger/web/controllers/AuthenticationController.java [hunk 2]
```diff
@@ -28,4 +33,15 @@ public AuthenticationController(HandlerFactory handlerFactory) {
...
+
+    @RequestMapping(value = "/register", method = RequestMethod.POST)
+    public ResponseEntity<SignUpResponse> register(@Valid @RequestBody SignUpRequest signUpRequest) {
+        SignUpHandler handler =
+                (SignUpHandler) handlerFactory.getHandler(HandlerFactory.SIGN_UP_HANDLER);
+
+        SignUpResponse response = handler.handle(signUpRequest);
+        return response.hasErrors()
+                ? ResponseEntity.badRequest().body(response)
+                : ResponseEntity.created(response.getCreatedResourceUri()).build();
```
Comment (VadymPolyanski) [inline src/main/java/com/george/messenger/web/controllers/AuthenticationController.java]: you can replace your 'errors' by RuntimeExceptions..with @ResponseStatus

File: src/main/java/com/george/messenger/web/handler/HandlerFactory.java [hunk 3]
```diff
@@ -31,6 +32,8 @@ public void init() {
 
             if (handler instanceof LoginHandler) {
                 key = LOGIN_HANDLER;
+            } else if (handler instanceof SignUpHandler){
+                key = SIGN_UP_HANDLER;
             } else {
```
Comment (VadymPolyanski) [inline src/main/java/com/george/messenger/web/handler/HandlerFactory.java]: replace this by Map<Class, Handler>...I think it'll be clearer

File: src/main/java/com/george/messenger/web/handler/SignUpHandler.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+                .setPassword(encoder.encode(request.getPassword()))
+                .setFirstName(request.getFirstName())
+                .setLastName(request.getLastName())
+                .setGender(request.getGender().equalsIgnoreCase("m") ? Gender.MALE : Gender.FEMALE)
+                .setAge(request.getAge());
+
+        userEntity = userService.save(userEntity);
+
+        return new SignUpResponse().setCreatedResourceUri(
+                buildURI(userEntity.getId())
```
Comment (VadymPolyanski) [inline src/main/java/com/george/messenger/web/handler/SignUpHandler.java]: create token for new user after registration

File: src/main/java/com/george/messenger/web/validation/SignUpRequestValidator.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+            errors.rejectValue("username", "username.taken", "Username is already taken");
+        }
+        if (request.getUsername().length() < 6) {
+            errors.rejectValue("username", "username.size", "Username field should be at least 6 characters");
+        }
+        if (request.getPassword().length() < 8 || request.getPassword().length() > 32) {
+            errors.rejectValue("password", "password.size", "Password length should be between 8 and 32 symbols");
+        }
+        if (request.getConfirmPassword().length() < 8 || request.getConfirmPassword().length() > 32) {
+            errors.rejectValue("confirmPassword", "password.size", "Password length should be between 8 and 32 symbols");
```
Comment (VadymPolyanski) [inline src/main/java/com/george/messenger/web/validation/SignUpRequestValidator.java]: you can use validation annotations in your request
1)http://docs.oracle.com/javaee/6/tutorial/doc/gircz.html
2)https://access.redhat.com/documentation/en-US/JBoss_Enterprise_Web_Platform/5/html-single/Hibernate_Validator_Reference_Guide/index.html

File: src/main/java/com/george/messenger/web/response/SignUpResponse.java (hunk 1 above)
Comment (edwardGr) [reply src/main/java/com/george/messenger/web/response/SignUpResponse.java]: Oh, right. So user wouldn't need to login after registration.

File: src/main/java/com/george/messenger/web/handler/HandlerFactory.java (hunk 3 above)
Comment (edwardGr) [reply src/main/java/com/george/messenger/web/handler/HandlerFactory.java]: Gotcha

File: src/main/java/com/george/messenger/web/validation/SignUpRequestValidator.java (hunk 5 above)
Comment (edwardGr) [reply src/main/java/com/george/messenger/web/validation/SignUpRequestValidator.java]: Okay. Thanks for provided links.

File: src/main/java/com/george/messenger/web/response/SignUpResponse.java (hunk 1 above)
Comment (VadymPolyanski) [reply src/main/java/com/george/messenger/web/response/SignUpResponse.java]: exactly

File: src/main/java/com/george/messenger/web/handler/LoginHandler.java [hunk 6]
```diff
@@ -39,7 +40,7 @@ public LoginResponse handle(LoginRequest request) {
         securityService.autoLogin(username, password);
         String token = tokenUtil.generateToken(userEntity.getId());
 
-        return new LoginResponse(token);
+        return ResponseEntity.ok().body(new LoginResponse(token));
     }
```
Comment (VadymPolyanski) [inline src/main/java/com/george/messenger/web/handler/LoginHandler.java]: good job!

File: src/main/java/com/george/messenger/web/handler/SignUpHandler.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+                .setUsername(request.getUsername())
+                .setPassword(encoder.encode(request.getPassword()))
+                .setFirstName(request.getFirstName())
+                .setLastName(request.getLastName())
+                .setGender(Gender.valueOf(request.getGender()))
+                .setAge(request.getAge());
+
+        userEntity = userService.save(userEntity);
+        URI createdResourceURI = buildURI(userEntity.getId());
+
```
Comment (VadymPolyanski) [inline src/main/java/com/george/messenger/web/handler/SignUpHandler.java]: return only token
don't build any uri

File: src/main/java/com/george/messenger/web/request/SignUpRequest.java [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    private String gender;
+
+    @NotNull(message = "{field.required}")
+    @Range(min = 16, max = 120, message = "{age.size}")
+    private Integer age;
+
+    @AssertTrue
+    public boolean isValid() {
+        return confirmPassword.equalsIgnoreCase(password);
+    }
```
Comment (VadymPolyanski) [inline src/main/java/com/george/messenger/web/request/SignUpRequest.java]: great

File: src/main/java/com/george/messenger/web/validation/SignUpRequestValidator.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+import com.george.messenger.domain.services.UserService;
+import com.george.messenger.web.request.SignUpRequest;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.validation.Errors;
+import org.springframework.validation.ValidationUtils;
+import org.springframework.validation.Validator;
+
+public class SignUpRequestValidator implements Validator {
+
```
Comment (VadymPolyanski) [inline src/main/java/com/george/messenger/web/validation/SignUpRequestValidator.java]: ???

File: src/test/java/com/george/messenger/domain/services/MessageServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+
+import java.util.Arrays;
+import java.util.Date;
+import java.util.List;
+
+import static org.junit.Assert.assertTrue;
+
+@ContextConfiguration
+@RunWith(SpringJUnit4ClassRunner.class)
+public class MessageServiceTest {
```
Comment (VadymPolyanski) [inline src/test/java/com/george/messenger/domain/services/MessageServiceTest.java]: mockito!!!

File: src/test/java/com/george/messenger/domain/services/MessageServiceTest.java [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+        @Primary
+        public MessageService messageService(MessageEntityRepository repository) {
+            return new MessageServiceImpl(repository);
+        }
+
+        @Bean
+        @Primary
+        public MessageEntityRepository messageEntityRepository() {
+            return Mockito.mock(MessageEntityRepository.class);
+        }
```
Comment (VadymPolyanski) [inline src/test/java/com/george/messenger/domain/services/MessageServiceTest.java]: mock repository and another services
it should return only expected values
test repository in repository tests

File: src/test/java/com/george/messenger/domain/services/UserServiceTest.java [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+        @Autowired
+        @Primary
+        public UserService userService(UserEntityRepository entityRepository) {
+            return new UserServiceImpl(entityRepository);
+        }
+
+        @Bean
+        @Primary
+        public UserEntityRepository entityRepository() {
+            return Mockito.mock(UserEntityRepository.class);
```
Comment (VadymPolyanski) [inline src/test/java/com/george/messenger/domain/services/UserServiceTest.java]: same

File: src/test/java/com/george/messenger/util/JwtTokenUtilTest.java [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+import static org.junit.Assert.assertEquals;
+
+public class JwtTokenUtilTest {
+
+    public static String SECRET = "secret";
+    public static long EXPIRATION = 604_800_000L;
+    public static String TEST_USER_ID = UUID.randomUUID().toString();
+
+    @InjectMocks
+    private JwtTokenUtil jwtTokenUtil;
```
Comment (VadymPolyanski) [inline src/test/java/com/george/messenger/util/JwtTokenUtilTest.java]: why do you use mockito here?

File: src/main/java/com/george/messenger/web/controllers/AuthenticationController.java [hunk 14]
```diff
@@ -24,8 +30,23 @@ public AuthenticationController(HandlerFactory handlerFactory) {
...
+
+        return handler.handle(loginRequest);
+    }
+
+    @RequestMapping(value = "/register", method = RequestMethod.POST)
+    public ResponseEntity<SignUpResponse> register(@Valid @RequestBody SignUpRequest signUpRequest) {
+        SignUpHandler handler =
+                (SignUpHandler) handlerFactory.getHandler(SignUpHandler.class);
+
+        return handler.handle(signUpRequest);
```
[the remark this answers is not in the document]
Comment (edwardGr) [reply src/main/java/com/george/messenger/web/controllers/AuthenticationController.java]: Yeah, I've done exactly what you've wrote but in different branch, gonna figure out how to merge these changes.

File: src/main/java/com/george/messenger/web/handler/SignUpHandler.java (hunk 7 above)
Comment (edwardGr) [reply src/main/java/com/george/messenger/web/handler/SignUpHandler.java]: Got it.

File: src/main/java/com/george/messenger/web/validation/SignUpRequestValidator.java (hunk 9 above)
Comment (edwardGr) [reply src/main/java/com/george/messenger/web/validation/SignUpRequestValidator.java]: Not really sure how this file got here. Definitely gonna remove it.

File: src/test/java/com/george/messenger/util/JwtTokenUtilTest.java (hunk 13 above)
Comment (edwardGr) [reply src/test/java/com/george/messenger/util/JwtTokenUtilTest.java]: Yeah, it doesn't make any sense. Headed to fix it.

File: src/main/java/com/george/messenger/web/controllers/AuthenticationController.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (VadymPolyanski) [reply src/main/java/com/george/messenger/web/controllers/AuthenticationController.java]: if this changes are merged to master you can rebase this branch(on master)

**Comment by VadymPolyanski** [comment]
59 files? are you joking?)
separate this on various PR

**Comment by edwardGr** [comment]
Should I revert this branch to the moment I commited last time. Then move those commits into separate branch and open PR?

**Comment by VadymPolyanski** [comment]
you cancreate new branch and move your commits by cherry pick

**Comment by edwardGr** [comment]
Hm, I can't quite comprehend what am I supposed to do.
I need to create new branch then cherry pick range of commits (like half of my new commits)?
Cause when I do so, half of code is broken.

**Comment by VadymPolyanski** [comment]
https://stackoverflow.com/questions/30768148/git-split-pull-request-into-smaller-prs-based-upon-the-new-directories-in-the

**Comment by VadymPolyanski** [comment]
remember that you have google ;)

**Comment by edwardGr** [comment]
Cherry-picked commits, that was necessary for this branch. So now I will create another branch with other changes.

**Pull request merged**